infra/network/src/main/java/com/simprints/infra/network/httpclient/DefaultOkHttpClientBuilder.kt

```diff
@@ -65,11 +72,55 @@ internal class DefaultOkHttpClientBuilder {
         }
     }
 
-    private fun buildVersionInterceptor(versionName: String): Interceptor =
-        Interceptor { chain ->
-            val newRequest = chain.request().newBuilder()
-                .addHeader(USER_AGENT_HEADER, "SimprintsID/$versionName")
-                .build()
-            return@Interceptor chain.proceed(newRequest)
+    private fun buildVersionInterceptor(versionName: String) = Interceptor { chain ->
+        val newRequest = chain.request().newBuilder()
+            .addHeader(USER_AGENT_HEADER, "SimprintsID/$versionName")
+            .build()
+        return@Interceptor chain.proceed(newRequest)
+    }
+
+    private fun buildGZipInterceptor() = Interceptor { chain ->
+        val originalRequest = chain.request()
+
+        // Only compress requests with explicit gzip header and non-null body
+        if (originalRequest.body == null || originalRequest.header("Content-Encoding") != "gzip") {
+            return@Interceptor chain.proceed(originalRequest)
         }
+
+        val compressedRequest = originalRequest.newBuilder()
+            .method(originalRequest.method, originalRequest.body?.let { gzip(it) })
+            .build()
+        return@Interceptor chain.proceed(compressedRequest)
+    }
+
+    // First compress the original RequestBody, then wrap it with a new RequestBody to set correct content length
+    // https://github.com/square/okhttp/issues/350
+    @Throws(IOException::class)
+    private fun gzip(requestBody: RequestBody): RequestBody {
+        val buffer = Buffer()
+        CompressedRequestBody(requestBody).writeTo(buffer)
+
+        return object : RequestBody() {
+            override fun contentType() = requestBody.contentType()
+
+            override fun contentLength() = buffer.size
+
+            @Throws(IOException::class)
+            override fun writeTo(sink: BufferedSink) {
+                sink.write(buffer.snapshot())
+            }
+        }
+    }
+
+    private class CompressedRequestBody(private val requestBody: RequestBody) : RequestBody() {
+
+        override fun contentType(): MediaType? = requestBody.contentType()
+
+        override fun contentLength(): Long = -1 // We don't know the compressed length in advance!
+
+        @Throws(IOException::class)
+        override fun writeTo(sink: BufferedSink) {
+            GzipSink(sink).buffer().use { gzipSink -> requestBody.writeTo(gzipSink) }
```

Comment:
Do you have any idea what compression level this is using by default?

Reply:
It is either none or some auto-detecting one

<img width="423" alt="Screenshot 2024-01-15 at 09 19 42" src="https://github.com/Simprints/Android-Simprints-ID/assets/2182889/e41c0182-6f01-4a2b-9502-ca8820260657">

Reply:
Explicitly set the compression level to 6.